src/util/psbt/mod.rs

```diff
@@ -0,0 +1,548 @@
+// Rust Bitcoin Library
+// Written by
+//   The Rust Bitcoin developers
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! # Partially Signed Transactions
+//!
+//! Implementation of BIP174 Partially Signed Bitcoin Transaction Format as
+//! defined at https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki
+//! except we define PSBTs containing non-standard SigHash types as invalid.
+
+use blockdata::transaction::Transaction;
+use consensus::encode::{self, Encodable, Decodable, Encoder, Decoder};
+
+mod error;
+pub use self::error::Error;
+
+pub mod raw;
+
+#[macro_use]
+mod macros;
+
+pub mod serialize;
+
+mod map;
+pub use self::map::{Map, Global, Input, Output};
+
+/// A Partially Signed Transaction.
+#[derive(Debug, Clone, PartialEq)]
+pub struct PartiallySignedTransaction {
+    /// The key-value pairs for all global data.
+    global: Global,
+    /// The corresponding key-value map for each input in the unsigned
+    /// transaction.
+    inputs: Vec<Input>,
+    /// The corresponding key-value map for each output in the unsigned
+    /// transaction.
+    outputs: Vec<Output>,
+}
+
+impl PartiallySignedTransaction {
+    /// Create a PartiallySignedTransaction from an unsigned transaction, error
+    /// if not unsigned
+    pub fn from_unsigned_tx(tx: Transaction) -> Result<Self, encode::Error> {
+        Ok(PartiallySignedTransaction {
+            inputs: vec![Default::default(); tx.input.len()],
+            outputs: vec![Default::default(); tx.output.len()],
+            global: Global::from_unsigned_tx(tx)?,
+        })
+    }
+
+    /// Extract a Transaction from a finalized PartiallySignedTransaction
+    pub fn extract_tx(self) -> Transaction {
```

Comment:
There doesn't seem to be a convenient way to know if a tx is "final". So this method not returning a `Result` is really annoying.

I think its really worth it to at least return an Option or so to catch the case in which the transaction is not finalized.

Another option would be to have an `is_finalized` inspector that can be called first.

Reply:
It's nontrivial to tell if a tx is final in general, without script descriptor support or having the bitcoinconsensus feature on.

Reply:
Well, this method returns a tx if all the "final signature" fields are filled. Of course that's no guarantee that the tx is valid etc. With this method returning an error, at least you know that something was missing without crashing your program :D

Reply:
Agreed this method should definitely not panic. I think these `unwrap()`s should be `unwrap_or(vec![])` or something. Having a nonempty final_script_sig or final_script_witness does not have much bearing on whether or not the transaction is final.